From 486323bed776800870f293821955ede146bfa6f6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Petr=20P=C3=ADsa=C5=99?= <ppisar@redhat.com>
Date: Tue, 12 Mar 2013 10:25:28 +0100
Subject: [PATCH] =?UTF-8?q?Don=E2=80=99t=20crash=20with=20formats=20in=20s?=
 =?UTF-8?q?pecial=20blocks?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Port to 5.10.1:

commit 4a273b91c8e47ab37c6dd310072403d4fe2d0fb9
Author: Father Chrysostomos <sprout@cpan.org>
Date:   Fri Jun 29 12:41:21 2012 -0700

    Don’t crash with formats in special blocks

    Commit 421f30ed1e9 didn’t go far enough.  If a special block happens
    to replace a stub, then a format trying to close over variables in the
    special block will be pointing to the wrong outer sub.

    Such stubs shouldn’t usually happen, but perl shouldn’t crash.

<https://bugzilla.redhat.com/show_bug.cgi?id=920132>
---
 perly.y             |  2 +-
 t/comp/form_scope.t | 15 ++++++++++++++-
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/perly.y b/perly.y
index 9164cab..96d7ff9 100644
--- a/perly.y
+++ b/perly.y
@@ -518,7 +518,7 @@ format	:	FORMAT startformsub formname block
 			  newFORM($2, $3, $4);
 			  $$ = (OP*)NULL;
 #endif
-			  if (CvOUTSIDE(fmtcv) && !CvUNIQUE(CvOUTSIDE(fmtcv))) {
+			  if (CvOUTSIDE(fmtcv) && !CvEVAL(CvOUTSIDE(fmtcv))) {
 			    SvREFCNT_inc_simple_void(fmtcv);
 			    pad_add_anon((SV*)fmtcv, OP_NULL);
 			  }
diff --git a/t/comp/form_scope.t b/t/comp/form_scope.t
index c0654d8..90244bf 100644
--- a/t/comp/form_scope.t
+++ b/t/comp/form_scope.t
@@ -1,6 +1,6 @@
 #!./perl
 
-print "1..7\n";
+print "1..8\n";
  
 # Tests bug #22977.  Test case from Dave Mitchell.
 sub f ($);
@@ -95,3 +95,16 @@ undef &x;
   print "ok 7 - closure var not available when outer sub is undefined\n";
 }
 
+# This is a variation of bug #22977, which crashes or fails an assertion
+# up to 5.16.
+# Keep this test last if you want test numbers to be sane.
+BEGIN { \&END }
+END {
+  my $test = "ok 8";
+  *STDOUT = *STDOUT5{FORMAT};
+  write;
+  format STDOUT5 =
+@<<<<<<<
+$test
+.
+}
-- 
1.8.1.4

